pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/multithreading/UnsynchronizedStaticMessageDigestRule.java

```diff
@@ -0,0 +1,19 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.rule.multithreading;
+
+import java.security.MessageDigest;
+
+/**
+ * Using a MessageDigest which is static can cause
+ * unexpected results when used in a multi-threaded environment. This rule will
+ * find static MessageDigest which are used in an unsynchronized manner.
+ */
+public class UnsynchronizedStaticMessageDigestRule extends UnsynchronizedStaticFormatterRule {
```

Comment:
having this extend `UnsynchronizedStaticFormatterRule` seems semantically incorrect even if it works… we should probably refactor `UnsynchronizedStaticFormatterRule` into a `UnsynchronizedStaticAccessRule` which can be configured through properties to track unsynchronized static access to any given types (with proper defaults to include `Formatter` and `MessageDigest`).

The `UnsynchronizedStaticFormatterRule` can be marked as deprecated at the ruleset level, point to the new implementation, overriding the property defaults to keep just looking for `Formatter`.

Moreover, the `THREAD_SAFE_FORMATTER` static should probably be replaced with a property of thread-safe classes to ignore, once again with proper defaults.